File: fixtures/sample-repos/playwright/support/factories.ts
// Factory: external test data only. No hardcoded strings, IDs, or
// credentials in test bodies. Each call returns a fresh object.
export interface User {
  email: string;
  password: string;
  role: 'trader' | 'admin' | 'viewer';
}

export const UserFactory = {
  create(overrides: Partial<User> = {}): User {
    const stamp = Date.now().toString(36);
    return {
      email: `user-${stamp}@example.test`,
      password: `pw-${stamp}`,
      role: 'trader',
      ...overrides,
    };
  },
};

File: fixtures/sample-repos/selenium-dotnet/Support/Factories.cs
using System;
using OpenQA.Selenium;
using OpenQA.Selenium.Chrome;

namespace Sample.SeleniumDotnet.Support;

public enum UserRole
{
    Trader,
    Admin,
    Viewer,
}

public record User(string Email, string Password, UserRole Role);

public static class UserFactory
{
    public static User Create(UserRole role = UserRole.Trader)
    {
        var stamp = DateTime.UtcNow.Ticks.ToString("x");
        return new User($"user-{stamp}@example.test", $"pw-{stamp}", role);
    }
}

public static class DriverFactory
{
    public static IWebDriver CreateChromeDriver()
    {
        var options = new ChromeOptions();
        options.AddArgument("--headless=new");
        return new ChromeDriver(options);
    }

    public static string BaseUrl =>
        Environment.GetEnvironmentVariable("SELENIUM_BASE_URL") ?? "http://127.0.0.1:5173";
}
